### PR TITLE
bpo-46075: Allow for explicit domains in CookieJar

### PR DESCRIPTION
**Pull request by keddad**
https://bugs.python.org/issue46075

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@keddad

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by kraktus** [comment]
@keddad gentle bump 👀

File: Lib/test/test_http_cookiejar.py [hunk 1]
```diff
@@ -920,6 +920,13 @@ def test_two_component_domain_ns(self):
 ##         self.assertEqual(len(c), 2)
         self.assertEqual(len(c), 4)
 
+    def test_localhost_domain(self):
+        c = CookieJar()
+
+        interact_netscape(c, "http://localhost", "foo=bar; domain=localhost;")
+
+        self.assertEqual(len(c), 1)
```
Comment (JelleZijlstra) [inline Lib/test/test_http_cookiejar.py:928]: Can you also test that the cookie in the jar is what you expect it to be?

File: Lib/http/cookiejar.py [hunk 2]
```diff
@@ -1044,12 +1044,13 @@ def set_ok_domain(self, cookie, request):
             else:
                 undotted_domain = domain
             embedded_dots = (undotted_domain.find(".") >= 0)
-            if not embedded_dots and domain != ".local":
+            if not embedded_dots and not erhn.endswith(".local"):
```
Comment (JelleZijlstra) [inline Lib/http/cookiejar.py:1047]: Previously this checked a domain that came from the cookie, and now it checks the hostname from the request. Can you explain why this is changed and why the change is safe?

File: Lib/http/cookiejar.py [hunk 3]
```diff
@@ -1044,12 +1044,13 @@ def set_ok_domain(self, cookie, request):
...
                 _debug("   non-local domain %s contains no embedded dot",
                        domain)
                 return False
             if cookie.version == 0:
-                if (not erhn.endswith(domain) and
+                if (not (erhn.endswith(domain) or
+                         (erhn.endswith(".local") and erhn[:len(erhn)-6].endswith(undotted_domain))) and
```
Comment (JelleZijlstra) [inline Lib/http/cookiejar.py]:
```suggestion
                         erhn.endswith(f"{undotted_domain}.local")) and
```
This is equivalent, right?

**Review by JelleZijlstra — COMMENTED** [review]
A few comments. I want to be very careful here because I'm not that familiar with cookie processing and I feel like getting something wrong here is a CVE waiting to happen.

File: Lib/http/cookiejar.py (hunk 2 above)
Comment (keddad) [reply Lib/http/cookiejar.py:1047]: This check considers `localhost` to be a non-local domain. It also doesn't have any embedded_dots, so it doesn't process it, which really shouldn't happen. My change should be safe because for local domains effective request hostname would end with ".local" (ERHN will end with .local if it ls clearly specified as part of the domain (`foo.local`), or if the domain consists of only one level (`localhost` has EHRN of `localhost.local`)).

File: Lib/http/cookiejar.py (hunk 3 above)
Comment (keddad) [reply Lib/http/cookiejar.py]: Yes, it definitely is, I'll commit it

File: Lib/test/test_http_cookiejar.py (hunk 1 above)
Comment (keddad) [reply Lib/test/test_http_cookiejar.py:928]: I've written such a test, and it looks like it saves the cookie under ".localhost". Without "domain" attribute present, it saves them under "localhost.local". I'm not so familiar with cookie processing as well, so I'm not really sure if "domain" attribute should affect this in such a way. I've commited the tests in [cec78ed](https://github.com/python/cpython/pull/30108/commits/cec78ed3eae6c648328338aefe4f47c8b59e6c54).

File: Lib/http/cookiejar.py (hunk 2 above)
Comment (JelleZijlstra) [reply Lib/http/cookiejar.py:1047]: Thanks! I'm still worried about cases where the domain and the ERHN don't match (e.g. a malicious website `evil.com` sets a cookie for `localhost`). I see there are some checks for that and a test in `test_wrong_domain`, but it's not very thorough. Could you add more tests for similar cases (request to non-local domain setting a local cookie, and vice versa)?

File: Lib/test/test_http_cookiejar.py (hunk 1 above)
Comment (JelleZijlstra) [reply Lib/test/test_http_cookiejar.py:928]: Yes, that's a bit odd. It might be fine though as long as it behaves consistently. The public interface doesn't expose the domain names as far as I can see.

File: Lib/http/cookiejar.py (hunk 2 above)
Comment (keddad) [reply Lib/http/cookiejar.py:1047]: I've added some tests for those cases in [a31fa40](https://github.com/python/cpython/pull/30108/commits/a31fa402ebf805692cc991f04751067d66b46b38). Looks like everything is fine.

File: Misc/NEWS.d/next/Library/2021-12-14-21-19-04.bpo-46075.KDtcU-.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+``CookieJar`` with ``DefaultCookiePolicy`` now can process cookies from localhost with domain=localhost explicitly specified in Set-Cookie header.
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Library/2021-12-14-21-19-04.bpo-46075.KDtcU-.rst:1]: CI is failing because there's no trailing newline. I tried to edit it from the web interface but wasn't successful, could you add it?

Comment (keddad) [reply Misc/NEWS.d/next/Library/2021-12-14-21-19-04.bpo-46075.KDtcU-.rst:1]: It looks like you added it? The CI passes.

Comment (JelleZijlstra) [reply Misc/NEWS.d/next/Library/2021-12-14-21-19-04.bpo-46075.KDtcU-.rst:1]: Ah right, I refreshed the page and no commit from me showed up, but maybe GH was just lagging.

**Review by JelleZijlstra — APPROVED** [review]
